autoload/thesaurus_query/thesaurus_query.py

```diff
@@ -61,47 +68,65 @@ def session_terminate(self):
             self.session_init()  # start a session if not started
-        # word not found, start searching
-        error_encountered = 0
-# use session-wise backend management to prepare for current query
+
+        found = False
+        local_bad_backends=[]
+        antonym_list=[]
+        synonym_list=[]
+
+        # Check cache first
+        if not self.cached_used and cache_results > -1:
+            current_cache = self.synonym_cache if query_type == 0 else self.antonym_cache
+            for cache_result in current_cache:
+                if cache_result[0] == word:
+                    self.cached_used = True
+                    self.good_backends = [cache_result[2] + ' (cached)']
+                    return cache_result[1]
+        # use session-wise backend management to prepare for current query
         if next:
             to_use_list = self.backend_in_line[:]
             success_list = self.good_backends[:]
             if len(to_use_list)==0:
-                return self.last_valid_result
+                return self.last_valid_synonyms if query_type == 0 else self.last_valid_antonyms
         else:
             to_use_list = self.good_backends[::-1]
             success_list = self.backend_in_line[::-1]
             if len(to_use_list)<=1:
-                return self.last_valid_result
+                return self.last_valid_synonyms if query_type == 0 else self.last_valid_antonyms
             success_list.append(to_use_list.pop(0))
-        local_bad_backends=[]
         for query_backend_curr in to_use_list:  # query each of the backend list till found
-            specified_language = get_variable("tq_language", ['en'])
-            if specified_language!="All" and \
-                    (self.query_backends[query_backend_curr].language not in \
-                        specified_language):
+            query_backend = self.query_backends.get(query_backend_curr, None)
+            if not query_backend:
+                continue
+            if specified_language!="All" and query_backend.language not in specified_language: 
                 continue
-            [state, synonym_list]=self.query_backends[query_backend_curr].query(word)
+            query_result = query_backend.query(word)
+            if (len(query_result) >= 3):
+                [state, synonym_list, antonym_list] = query_result
+            else:
+                [state, synonym_list] = query_result
+                antonym_list = []
+                if query_type == 1:
+                    state = 1
             if state == -1:
-                error_encountered = 1
-                local_bad_backends.append(
-                    self.query_backends[query_backend_curr].identifier)
+                local_bad_backends.append(query_backend.identifier)
                 continue
             if state == 0:
+                # Update caches
+                update_cache(self.antonym_cache, word, antonym_list, query_backend.identifier)
```

Comment:
since our `state` determination is based on the type of related word we try to search. Shouldn't our cache saving also based on our current query? Or else if the current search is on synonym from a backend that does not support antonym, then it will generate an empty cache for the antonym here.
Let's do
```
if state==0:
    if query_type==0:
        update_cache(self.antonym_cache, word, antonym_list, query_backend.identifier)
    else:
        update_cache(self.synonym_cache, word, synonym_list, query_backend.identifier)
```